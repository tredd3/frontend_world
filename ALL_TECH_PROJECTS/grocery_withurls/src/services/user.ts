import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { formatDate } from '../helpers/functional';
import { pivotAPIUser } from './helpers/pivots';
import { APIUser, UserPivot } from '../types';
import { genderStringToCode } from './helpers/generic';

export const getUser = promiseMemoize(async (): Promise<UserPivot> => {
  const user = await makeApiCall<{}, APIUser>(apiRoutes.getUserDetail);
  return pivotAPIUser(user);
});

type UpdateUserArgs = {
  userId?: number;
  firstName?: string;
  lastName?: string;
  email?: string;
  gender?: UserPivot['gender'];
  dob?: Date;
  lat?: number;
  lng?: number;
  pincode?: number;
  city?: string;
  storeId?: number;
  storeName?: string;
};

export const updateUser = async ({
  firstName, lastName, email, gender, dob, lat, lng, pincode, city, storeId, storeName
}: UpdateUserArgs) => {
  const userFromApi = await getUser();
  const { userId } = userFromApi;
  const isValidStoreId = storeId !== undefined;
  const isValidStoreName = storeName !== undefined;

  await makeApiCall(apiRoutes.updateUser, {
    body: {
      userId,
      ...(firstName && { firstName }),
      ...(lastName && { lastName }),
      ...(email ? { emailId: email } : { emailId: '' }),
      ...(gender && { gender: genderStringToCode(gender) }),
      ...(dob ? { dob: formatDate(dob) } : { dob: '' }),
      preferences: {
        ...userFromApi.preferences,
        ...(pincode && { pincode }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
        ...(isValidStoreId && { storeId }),
        ...(city && { city }),
        ...(isValidStoreName && { storeName })
      }
    }
  });

  getUser.clear();
};

type AddUserArgs = {
  firstName: string;
  lastName: string;
};

export const addUser = ({ firstName, lastName }: AddUserArgs) => (
  makeApiCall<AddUserArgs, void>(apiRoutes.addUser, {
    body: {
      firstName,
      lastName
    }
  })
);
