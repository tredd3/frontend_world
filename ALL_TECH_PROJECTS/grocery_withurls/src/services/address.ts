import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { getUser, updateUser } from './user';
import { APIBaseAddress, APIAddress, AddressPivot } from '../types';
import { pivotAPIAddress } from './helpers/pivots';

export type APIAddressRequest = APIBaseAddress & {
  addressId: number;
  addedDate: string;
};

// eslint-disable-next-line camelcase, @typescript-eslint/camelcase
export const DEPRECATED_convertAddressToAPIAddress = ({ id, addedDate, ...address }: APIAddress): APIAddressRequest => (
  { addressId: id, addedDate: addedDate.toDateString(), ...address }
);

type AreaType = { pincode: number};
export const isAreaServiceable = promiseMemoize(async ({ pincode }: AreaType) => {
  try {
    await makeApiCall<AreaType, {}>(apiRoutes.checkServiceability, { body: { pincode } });
    return true;
  } catch (e) {
    return false;
  }
}, { resolve: 'json' });

export const getAddresses = promiseMemoize(async (): Promise<AddressPivot[]> => {
  const { userId } = await getUser();
  const { addresses } = await makeApiCall<{ userId: number }, {addresses: APIAddressRequest[]}>(apiRoutes.getAddress, { body: { userId } });
  return addresses.map(pivotAPIAddress);
});

export const getAddressById = async (id: number): Promise<AddressPivot | undefined> => {
  const addresses = (await getAddresses()) || [];
  return addresses.find(address => address.id === id);
};

type UpdateAddressRequest = (APIAddressRequest | APIBaseAddress) & { userId?: number };

export const updateAddress = async (address: Omit<APIBaseAddress, 'address'> | APIAddress): Promise<APIAddress> => {
  const { id } = address as APIAddress;
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { addedDate, ...remainingAddress } = address as APIAddress;
  const newAddress = await makeApiCall<UpdateAddressRequest, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      ...(id && { addressId: id }),
      ...({ userId: (await getUser()).userId }),
      ...remainingAddress,
      address: `${remainingAddress.cityName}, ${remainingAddress.pincode}`
    }
  });

  getAddresses.clear();
  return pivotAPIAddress(newAddress);
};

export const addAddress = (address: Omit<APIBaseAddress, 'address'>) => updateAddress(address);

export const deleteAddress = async (id: number) => {
  getAddresses.clear();
  return makeApiCall<{ addressId: number }, void>(
    apiRoutes.deleteAddress, { body: { addressId: id } }
  );
};

export const setDefaultAddress = async (id: number) => {
  const address = await getAddressById(id);
  if (!address) throw new Error('An address with this id doesn\'t exist');

  await makeApiCall<{ addressId: number; isDefault: boolean }, APIAddressRequest>(apiRoutes.saveAddress, {
    body: {
      addressId: id,
      isDefault: true
    }
  });
  getAddresses.clear();

  await updateUser({
    lat: address.latitude,
    lng: address.longitude,
    pincode: address.pincode,
    storeId: address.storeId,
    storeName: address.storeName
  });

  return address;
};

export const getDefaultAddress = async () => {
  const addresses = await getAddresses();
  const defaultAddress = addresses.find(address => address.isDefault);
  if (defaultAddress) return defaultAddress;

  if (!addresses.length) return null;

  await setDefaultAddress(addresses[0].id);
  return addresses[0];
};

export const updateAddressWithKirana = async (addressId: number, { storeId, storeName }: { storeId: number; storeName: string }) => {
  const address = await getAddressById(addressId);

  if (!address) throw new Error('Couldn\'t find an address with this id');

  const updatedAddress = await updateAddress({ ...address, storeId, storeName });

  if (address.isDefault) {
    await updateUser({ storeId, storeName });
  }

  return updatedAddress;
};

export const deleteKiranaFromAddress = async (addressId: number) => updateAddressWithKirana(addressId, { storeId: 0, storeName: '' });

export const getStoreIdFromAddressId = async (addressId: number) => {
  const address = await getAddressById(addressId);

  if (!address) return null;

  return address.storeId || address.defaultStore || null;
};

export const getPincodeFromLatLng = async ({ lat, lng }: {lat: number; lng: number}): Promise<string> => {
  const geocoder = new window.google.maps.Geocoder();
  const pincode = new Promise<string>((resolve, reject) => {
    geocoder.geocode({ location: new window.google.maps.LatLng(lat, lng) }, (results, status) => {
      if (status === window.google.maps.GeocoderStatus.OK) {
        const len = results[0].address_components.length;
        resolve(results[0].address_components[len - 1].long_name);
      } else {
        reject();
      }
    });
  });
  return pincode;
};
