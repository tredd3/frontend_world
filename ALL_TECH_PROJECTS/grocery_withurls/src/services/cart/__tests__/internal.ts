import { mocked } from 'ts-jest/utils';
import {
  getCart, getUserIdStoreIdAndCartId, getDeliveryModes,
  deliveryDateByDateId, applyAddressChangeToCart, checkoutWithDeliveryDate
} from '../internal';
import mockDB from '../../../mocks/db.json';
import * as makeApiCall from '../../helpers/make-api-call';
import { apiRoutes } from '../../api-routes';
import { pivotAPIUser, pivotAPIAddress } from '../../helpers/pivots';
import * as user from '../../user';
import * as address from '../../address';

jest.mock('../../user');
jest.mock('../../address');
const { getUser } = mocked(user);
const { getStoreIdFromAddressId, getAddressById } = mocked(address);

describe('cart: internal', () => {
  let makeApiCallSpy: jest.SpyInstance<Promise<unknown>>;

  afterEach(() => {
    getCart.clear();
    getUser.mockClear();
    makeApiCallSpy.mockRestore();
    getDeliveryModes.clear();
    applyAddressChangeToCart.clear();
    checkoutWithDeliveryDate.clear();
  });

  describe('getCart', () => {
    it('should get the cart with the user\'s default store ID', async () => {
      getUser.mockImplementation(async () => pivotAPIUser(mockDB.user.Data));
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      const cart = await getCart();
      expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toEqual({
        body: { userId: (await getUser()).userId },
        headers: { StoreId: '0' }
      });
      expect(cart).toMatchSnapshot();
    });

    it('should get the cart with the user\'s default store ID', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 123123 }
      }));
      const { userId, preferences: { storeId } } = await getUser();
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      await getCart();

      expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toEqual({
        body: { userId, storeId },
        headers: { StoreId: String(storeId) }
      });
    });

    it('should get the cart with the specified store ID', async () => {
      getUser.mockImplementation(async () => pivotAPIUser(mockDB.user.Data));
      const { userId } = await getUser();
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      await getCart(321321);

      expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toEqual({
        body: { userId, storeId: 321321 },
        headers: { StoreId: '321321' }
      });
    });
  });

  describe('getUserIdStoreIdAndCartId', () => {
    it('should return store ID from user profile when addressId isn\'t passed in', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      const { userId } = await getUser();
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      const result = await getUserIdStoreIdAndCartId();

      expect(result).toEqual({ userId, storeId: 999999, cartId: mockDB.cart.Data.CartId });
      expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
    });

    it('should return store ID from user addresses when addressId is passed in', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      const { userId } = await getUser();
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);
      getStoreIdFromAddressId.mockImplementation(async () => 111111);

      const result = await getUserIdStoreIdAndCartId(123123);

      expect(result).toEqual({ userId, storeId: 111111, cartId: mockDB.cart.Data.CartId });
      expect(makeApiCallSpy).toHaveBeenCalledTimes(1);
    });

    it('should throw when addressId is passed in and there\'s store ID for that address ID', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);
      getStoreIdFromAddressId.mockImplementation(async () => null);
      expect.assertions(1);

      await expect(getUserIdStoreIdAndCartId(123123))
        .rejects
        .toBeInstanceOf(Error);
    });
  });

  describe('getDeliveryModes', () => {
    it('should get the delivery modes for an address ID', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      getAddressById.mockImplementation(async () => (
        pivotAPIAddress(mockDB.address.Data.addresses[0] as address.APIAddressRequest)
      ));
      getStoreIdFromAddressId.mockImplementation(async () => 123123);
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => mockDB.cart.Data)
        .mockImplementationOnce(async () => mockDB.modes.Data);

      const result = await getDeliveryModes(111111);

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1].headers).toEqual({ StoreId: '999999' });
      expect(makeApiCallSpy.mock.calls[1][0]).toBe(apiRoutes.getDeliveryModes);
      expect(makeApiCallSpy.mock.calls[1][1]).toEqual({
        body: {
          OrderValue: mockDB.cart.Data.TotalPay,
          Pincode: mockDB.address.Data.addresses[0].pincode,
          StoreId: 123123
        },
        headers: {
          StoreId: '123123'
        }
      });
      expect(result).toMatchSnapshot();
    });
  });

  describe('deliveryDateByDateId', () => {
    it('should return delivery dates for a date id', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      getAddressById.mockImplementation(async () => (
        pivotAPIAddress(mockDB.address.Data.addresses[0] as address.APIAddressRequest)
      ));
      getStoreIdFromAddressId.mockImplementation(async () => 123123);
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => mockDB.cart.Data)
        .mockImplementationOnce(async () => mockDB.modes.Data);

      const result = await deliveryDateByDateId(111111, mockDB.modes.Data.StoreOrder.StorePickup[0].DateId);

      expect(makeApiCallSpy.mock.calls[1][0]).toBe(apiRoutes.getDeliveryModes);
      expect(makeApiCallSpy.mock.calls[1][1].headers).toEqual({ StoreId: '123123' });
      expect(result).toBe(mockDB.modes.Data.StoreOrder.StorePickup[0].DateValue);
    });
  });

  describe('applyAddressChangeToCart', () => {
    it('should throw when address doesn\'t have a store ID', async () => {
      getStoreIdFromAddressId.mockImplementation(async () => null);

      await expect(applyAddressChangeToCart(123123))
        .rejects
        .toBeInstanceOf(Error);
    });

    it('should set the cartHasBeenUpdated flag to false if there\'s no cart change', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      getStoreIdFromAddressId.mockImplementation(async () => 123123);
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      const result = await applyAddressChangeToCart(999999);
      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toEqual({
        body: { storeId: 999999, userId: (await getUser()).userId },
        headers: { StoreId: '999999' }
      });
      expect(makeApiCallSpy.mock.calls[1][1]).toEqual({
        body: { storeId: 123123, userId: (await getUser()).userId },
        headers: { StoreId: '123123' }
      });
      expect(result.cartHasBeenUpdated).toBe(false);
      expect(result.oldCart).toEqual(result.newCart);
    });

    it('should set the cartHasBeenUpdated flag to true if there\'s a cart change', async () => {
      getUser.mockImplementation(async () => ({
        ...pivotAPIUser(mockDB.user.Data),
        preferences: { storeId: 999999 }
      }));
      getStoreIdFromAddressId.mockImplementation(async () => 123123);
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => mockDB.cart.Data)
        .mockImplementationOnce(async () => ({
          ...mockDB.cart.Data,
          TotalPay: mockDB.cart.Data.TotalPay + 10
        }));

      const result = await applyAddressChangeToCart(999999);

      expect(makeApiCallSpy.mock.calls[0][0]).toBe(apiRoutes.getCart);
      expect(makeApiCallSpy.mock.calls[0][1]).toEqual({
        body: { storeId: 999999, userId: (await getUser()).userId },
        headers: { StoreId: '999999' }
      });
      expect(makeApiCallSpy.mock.calls[1][1]).toEqual({
        body: { storeId: 123123, userId: (await getUser()).userId },
        headers: { StoreId: '123123' }
      });
      expect(result.cartHasBeenUpdated).toBe(true);
      expect(
        result.newCart.totalPayableAmount - result.oldCart.totalPayableAmount
      ).toEqual(10);
    });
  });

  describe('checkoutWithDeliveryDate', () => {
    it('should throw if it couldn\'t find a matching address', async () => {
      getUser.mockImplementation(async () => pivotAPIUser(mockDB.user.Data));
      getStoreIdFromAddressId.mockImplementation(async () => 111111); // Doesn't matter
      getAddressById.mockImplementation(async () => undefined);
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementation(async () => mockDB.cart.Data);

      await expect(checkoutWithDeliveryDate(123123, 1))
        .rejects
        .toBeInstanceOf(Error);
    });

    it('should call checkout correctly', async () => {
      getUser.mockImplementation(async () => pivotAPIUser(mockDB.user.Data));
      getStoreIdFromAddressId.mockImplementation(async () => 111111); // Doesn't matter
      getAddressById.mockImplementation(async () => (
        pivotAPIAddress(mockDB.address.Data.addresses[0] as address.APIAddressRequest)
      ));
      makeApiCallSpy = jest
        .spyOn(makeApiCall, 'default')
        .mockImplementationOnce(async () => mockDB.cart.Data)
        .mockImplementationOnce(async () => mockDB.modes.Data)
        .mockImplementationOnce(async () => mockDB.checkout.Data);

      const result = await checkoutWithDeliveryDate(123123, mockDB.modes.Data.StoreOrder.StorePickup[0].DateId);

      const getMockCalls = (path: string) => makeApiCallSpy.mock.calls
        .filter(([p]) => p === path);

      expect(getMockCalls(apiRoutes.getCart).length).toBe(1);
      expect(getMockCalls(apiRoutes.getCart)[0][1].headers).toEqual({ StoreId: '0' });
      expect(getMockCalls(apiRoutes.checkout)).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });
});
