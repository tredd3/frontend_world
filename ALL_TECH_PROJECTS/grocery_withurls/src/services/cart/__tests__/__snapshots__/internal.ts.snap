// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart: internal checkoutWithDeliveryDate should call checkout correctly 1`] = `
Array [
  Array [
    "/jiokart/v1/order/checkout",
    Object {
      "body": Object {
        "action": 1,
        "cartId": 2764873131,
        "customerAddressId": 10160497,
        "deliveryDate": "07-Jan-2020",
        "isAutoSelected": 0,
        "storeId": 111111,
        "timeslot": "",
        "userId": 10005564,
      },
      "headers": Object {
        "StoreId": "111111",
      },
    },
  ],
]
`;

exports[`cart: internal checkoutWithDeliveryDate should call checkout correctly 2`] = `
Object {
  "cartHasBeenUpdated": true,
  "newCart": Object {
    "cartId": "4062847169",
    "couponDiscountAmount": 0,
    "deliveryCharge": 0,
    "discountAmount": 11.399999999999991,
    "itemCount": 5,
    "savedForLater": null,
    "shipments": Array [
      Object {
        "couponDiscountAmount": 0,
        "creationDate": "06-Jan-2020",
        "deliveryCharge": 0,
        "deliveryDate": "",
        "deliveryDay": "Tomorrow",
        "discountAmount": 11.399999999999991,
        "itemCount": 5,
        "products": Array [
          Object {
            "categoryId": "10001310",
            "coupon": undefined,
            "id": "100023724",
            "image": "https://api.reliancesmart.in//images/?name=images/hc1/ha5/8920196907038.jpg",
            "isFc": 1,
            "mrp": 180,
            "name": "Mapro Lemon Ginger Squash",
            "otherImages": Array [
              "https://api.reliancesmart.in//images/?name=images/hc1/ha5/8920196907038.jpg",
              "https://api.reliancesmart.in//images/?name=images/hb7/h92/8920188256286.jpg",
              "https://api.reliancesmart.in//images/?name=images/h52/hd4/8920195661854.jpg",
              "https://api.reliancesmart.in//images/?name=images/h35/h18/8920201035806.jpg",
              "https://api.reliancesmart.in//images/?name=images/hf3/h1e/8920203722782.jpg",
              "https://api.reliancesmart.in//images/?name=images/h60/hb5/9021983817758.jpg",
            ],
            "savingsPercent": 2,
            "selectedQuantity": 1,
            "skuId": "10027751",
            "sp": 176.4,
          },
          Object {
            "categoryId": "10001098",
            "coupon": undefined,
            "id": "100030032",
            "image": "https://api.reliancesmart.in//images/?name=images/h16/h73/9031821754398.jpg",
            "isFc": 1,
            "mrp": 145,
            "name": "Glucon-D Tangy Orange 450 g",
            "otherImages": Array [
              "https://api.reliancesmart.in//images/?name=images/h16/h73/9031821754398.jpg",
              "https://api.reliancesmart.in//images/?name=images/he2/ha9/9031827062814.jpg",
              "https://api.reliancesmart.in//images/?name=images/h05/h13/9031834796062.jpg",
              "https://api.reliancesmart.in//images/?name=images/h02/h68/9031845412894.jpg",
              "https://api.reliancesmart.in//images/?name=images/h29/h08/9031849869342.jpg",
              "https://api.reliancesmart.in//images/?name=images/hcc/ha7/9031864549406.jpg",
              "https://api.reliancesmart.in//images/?name=images/h07/hdb/9031872610334.jpg",
              "https://api.reliancesmart.in//images/?name=images/hbe/h36/9032238104606.jpg",
              "https://api.reliancesmart.in//images/?name=images/h93/h06/9032248328222.jpg",
            ],
            "savingsPercent": 2,
            "selectedQuantity": 1,
            "skuId": "10042508",
            "sp": 142.1,
          },
          Object {
            "categoryId": "10001118",
            "coupon": undefined,
            "id": "100030059",
            "image": "https://api.reliancesmart.in//images/?name=images/hfa/hb3/9037997408286.jpg",
            "isFc": 1,
            "mrp": 135,
            "name": "Horlicks Junior Health Drink Original Flavour",
            "otherImages": Array [
              "https://api.reliancesmart.in//images/?name=images/hfa/hb3/9037997408286.jpg",
              "https://api.reliancesmart.in//images/?name=images/hc7/h0a/9037993476126.jpg",
              "https://api.reliancesmart.in//images/?name=images/h79/h26/9037994524702.jpg",
              "https://api.reliancesmart.in//images/?name=images/he0/h83/9037995507742.jpg",
              "https://api.reliancesmart.in//images/?name=images/h84/h76/9037996490782.jpg",
              "https://api.reliancesmart.in//images/?name=images/h03/h7b/9037997146142.jpg",
              "https://api.reliancesmart.in//images/?name=images/ha8/h13/9037997932574.jpg",
              "https://api.reliancesmart.in//images/?name=images/hd6/hf1/9037998587934.jpg",
            ],
            "savingsPercent": 2,
            "selectedQuantity": 1,
            "skuId": "10035638",
            "sp": 132.3,
          },
          Object {
            "categoryId": "10000994",
            "coupon": undefined,
            "id": "100005584",
            "image": "https://api.reliancesmart.in//images/?name=images/h3d/h6d/8996511612958.jpg",
            "isFc": 1,
            "mrp": 110,
            "name": "Tango Neapolitan Blueberry Chocolate Bar 110 g",
            "otherImages": Array [
              "https://api.reliancesmart.in//images/?name=images/h3d/h6d/8996511612958.jpg",
              "https://api.reliancesmart.in//images/?name=images/h2c/h8c/8996550213662.jpg",
              "https://api.reliancesmart.in//images/?name=images/h06/h6c/8996556963870.jpg",
              "https://api.reliancesmart.in//images/?name=images/hc8/h36/8996584095774.jpg",
            ],
            "savingsPercent": 2,
            "selectedQuantity": 1,
            "skuId": "10006440",
            "sp": 107.8,
          },
          Object {
            "categoryId": "10001870",
            "coupon": undefined,
            "id": "100094190",
            "image": "https://api.reliancesmart.in/",
            "isFc": 1,
            "mrp": 0.01,
            "name": "Welcome Gift For You",
            "otherImages": Array [
              "https://api.reliancesmart.in/",
            ],
            "savingsPercent": 0,
            "selectedQuantity": 1,
            "skuId": "10106839",
            "sp": 0.01,
          },
        ],
        "totalAmount": 558.61,
        "totalPayable": 558.61,
        "type": 0,
      },
    ],
    "totalAmount": 558.61,
    "totalPayableAmount": 558.61,
  },
  "oldCart": Object {
    "cartId": 2764873131,
    "couponDiscountAmount": 20,
    "deliveryCharge": 0,
    "discountAmount": 1.8999999999999986,
    "itemCount": 3,
    "savedForLater": null,
    "shipments": Array [
      Object {
        "couponDiscountAmount": 20,
        "creationDate": "17-Apr-2020",
        "deliveryCharge": 0,
        "deliveryDate": "",
        "deliveryDay": "Tomorrow",
        "discountAmount": 1.8999999999999986,
        "itemCount": 3,
        "products": Array [
          Object {
            "categoryId": 10002340,
            "coupon": null,
            "id": 100024066,
            "image": "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
            "isFc": 1,
            "mrp": 240,
            "name": "Nivea Smooth Body Milk",
            "otherImages": Array [
              "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/h16/h3a/8916957954078.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hc8/h0a/8916962082846.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hf9/h66/8916966604830.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hcf/hfa/8916976173086.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/h74/he3/8916983447582.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hf6/h4a/8916987183134.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/h56/hce/8917003763742.jpg",
            ],
            "savingsPercent": 0,
            "selectedQuantity": 1,
            "skuId": 10028095,
            "sp": 240,
          },
          Object {
            "categoryId": 10004128,
            "coupon": Object {
              "code": "",
              "description": "http://www.google.com",
              "discountAmount": 10,
              "id": 4170,
              "title": "10 rs off on Catch Whole Jeera 100 g",
            },
            "id": 100008042,
            "image": "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
            "isFc": 1,
            "mrp": 38,
            "name": "Catch Whole Jeera 100 g",
            "otherImages": Array [
              "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hc2/h06/8917068120094.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/hef/h25/8917070151710.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/h14/hee/8917072052254.jpg",
            ],
            "savingsPercent": 5,
            "selectedQuantity": 1,
            "skuId": 10009279,
            "sp": 36.1,
          },
          Object {
            "categoryId": 10003881,
            "coupon": Object {
              "code": "",
              "description": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
              "discountAmount": 10,
              "id": 4009,
              "title": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
            },
            "id": 100038166,
            "image": "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
            "isFc": 1,
            "mrp": 840,
            "name": "Whiskas Ocean Fish Adult Cat Food 3 kg",
            "otherImages": Array [
              "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
              "https://uat-grocery.ril.com:443//images/?name=images/h37/hba/8917010710558.jpg",
            ],
            "savingsPercent": 0,
            "selectedQuantity": 1,
            "skuId": 10043518,
            "sp": 840,
          },
        ],
        "totalAmount": 1116.1,
        "totalPayable": 1096.1,
        "type": 1,
      },
    ],
    "totalAmount": 1116.1,
    "totalPayableAmount": 1096.1,
  },
}
`;

exports[`cart: internal getCart should get the cart with the user's default store ID 1`] = `
Object {
  "cartId": 2764873131,
  "couponDiscountAmount": 20,
  "deliveryCharge": 0,
  "discountAmount": 1.8999999999999986,
  "itemCount": 3,
  "savedForLater": null,
  "shipments": Array [
    Object {
      "couponDiscountAmount": 20,
      "creationDate": "17-Apr-2020",
      "deliveryCharge": 0,
      "deliveryDate": "",
      "deliveryDay": "Tomorrow",
      "discountAmount": 1.8999999999999986,
      "itemCount": 3,
      "products": Array [
        Object {
          "categoryId": 10002340,
          "coupon": null,
          "id": 100024066,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
          "isFc": 1,
          "mrp": 240,
          "name": "Nivea Smooth Body Milk",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/hc1/hcb/8916991115294.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h16/h3a/8916957954078.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hc8/h0a/8916962082846.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hf9/h66/8916966604830.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hcf/hfa/8916976173086.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h74/he3/8916983447582.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hf6/h4a/8916987183134.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h56/hce/8917003763742.jpg",
          ],
          "savingsPercent": 0,
          "selectedQuantity": 1,
          "skuId": 10028095,
          "sp": 240,
        },
        Object {
          "categoryId": 10004128,
          "coupon": Object {
            "code": "",
            "description": "http://www.google.com",
            "discountAmount": 10,
            "id": 4170,
            "title": "10 rs off on Catch Whole Jeera 100 g",
          },
          "id": 100008042,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
          "isFc": 1,
          "mrp": 38,
          "name": "Catch Whole Jeera 100 g",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/h1c/h24/8917070282782.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hc2/h06/8917068120094.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/hef/h25/8917070151710.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h14/hee/8917072052254.jpg",
          ],
          "savingsPercent": 5,
          "selectedQuantity": 1,
          "skuId": 10009279,
          "sp": 36.1,
        },
        Object {
          "categoryId": 10003881,
          "coupon": Object {
            "code": "",
            "description": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
            "discountAmount": 10,
            "id": 4009,
            "title": "Rs 10 off on Whiskas Ocean Fish Adult Cat Food 3kg",
          },
          "id": 100038166,
          "image": "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
          "isFc": 1,
          "mrp": 840,
          "name": "Whiskas Ocean Fish Adult Cat Food 3 kg",
          "otherImages": Array [
            "https://uat-grocery.ril.com:443//images/?name=images/h64/he4/8916990001182.jpg",
            "https://uat-grocery.ril.com:443//images/?name=images/h37/hba/8917010710558.jpg",
          ],
          "savingsPercent": 0,
          "selectedQuantity": 1,
          "skuId": 10043518,
          "sp": 840,
        },
      ],
      "totalAmount": 1116.1,
      "totalPayable": 1096.1,
      "type": 1,
    },
  ],
  "totalAmount": 1116.1,
  "totalPayableAmount": 1096.1,
}
`;

exports[`cart: internal getDeliveryModes should get the delivery modes for an address ID 1`] = `
Array [
  Object {
    "id": 2,
    "isOpen": true,
    "label": "Tomorrow - 07 January 2020",
    "store": Object {
      "isKdp": false,
      "name": "Gupta Store",
      "storeId": 1000001,
    },
    "value": "07-Jan-2020",
  },
  Object {
    "id": 3,
    "isOpen": true,
    "label": "Wednesday - 08 January 2020",
    "store": Object {
      "isKdp": false,
      "name": "Gupta Store",
      "storeId": 1000001,
    },
    "value": "08-Jan-2020",
  },
  Object {
    "id": 4,
    "isOpen": true,
    "label": "Thursday - 09 January 2020",
    "store": Object {
      "isKdp": false,
      "name": "Gupta Store",
      "storeId": 1000001,
    },
    "value": "09-Jan-2020",
  },
]
`;
